### DISCUSSION TITLE
Launch captive portal when reconnect fails several times

### DISCUSSION THREAD

**Discussion by hjkoster400d** [Hieromon/AutoConnect#525] [Q&A] [answered]
Hi there,

Just another use case that i ran into:
1. ESP is connected to an AP (using AutoConnect)
2. Gets disconnected because the AP no longer exists or is renamed
3. AutoConnect will then try to reconnect for a few times

When it's not coming back online, i would like to restart the ESP and launch the captive portal and softAP. When nothing happens, reboot the ESP and try it all again until the network gets back up or someone enters the captive portal and sets it to use the 'new' network. 

Is this functionality possible with the current version?

**Comment by FRANAIRBUS** [comment]
Hello @hjkoster400d 

Good evening
If your requirement is possible, if you wish you can consult the documentation where your use case is explained.
If you have any questions, feel free to ask, but I think that will solve your request.
https://hieromon.github.io/AutoConnect/adcpcontrol.html
a greeting

**Comment by Hieromon** [comment, marked as answer]
@hjkoster400d

AutoConnect has the ability to attempt to reconnect in the background during a *AutoConnect::handleClient* loop after disconnecting from an AP. It is as @FRANAIRBUS mentioned, please refer to the [documentation](https://hieromon.github.io/AutoConnect/adconnection.html#automatic-reconnect-background) for details and configure the settings that match your use case. Perhaps [AutoConnectConfig::autoReconnect](https://hieromon.github.io/AutoConnect/apiconfig.html#autoreconnect), [AutoConnectConfig::portalTimeout](https://hieromon.github.io/AutoConnect/apiconfig.html#portaltimeout), [AutoConnectConfig::reconnectInterval](https://hieromon.github.io/AutoConnect/apiconfig.html#reconnectinterval), and [AutoConnectConfig::retainPortal](https://hieromon.github.io/AutoConnect/apiconfig.html#retainportal) apply. Also, I think the following code snippet will help you.

```cpp
AutoConnect portal;
AutoConnectConfig config;

void setup() {
  config.autoReconnect = true;  // Reconnect to known access points.
  config.portalTimeout = 3 * 60 * 1000; // Stay in Captive portal for 180 seconds.
  config.reconnectInterval = 6; // Reconnection attempting interval is 3[min] during handleClient().
  config.retainPortal = true;   // Keep the captive portal open.
  portal.config(config);
  portal.begin();
}

void loop() {
  if (WiFi.status() == WL_CONNECTED) {
    // Here to do when WiFi is connected.
  }
  else {
    // Here to do when WiFi is not connected.
  }
  portal.handleClient ();
}
```

However, a following item in your requirements may need to be considered more rigorously.
> When nothing happens, reboot the ESP and try it all again

There are various reasons why this may be the case, whether it is due to distance from the AP, loss of power to the AP, or the number of clients that can connect to the AP has exceeded the limit and has been kicked out. But it is a vague criterion for judgment; *"When nothing happens"*.

A common method is to measure the time elapsed while trying to reconnect automatically after a WiFi disconnection. You can detect the launch of the captive portal with the [onDetect](https://hieromon.github.io/AutoConnect/faq.html#how-can-i-detect-the-captive-portal-starting) exit and trigger the elapsed time measurement, then reboot due to a reconnect-timeout during the *loop()* function.

But what if the AP password has been changed? The *AutoConnect::begin* after reboot will always fail. Then the captive portal will start but will eventually exits from the *setup()* function due to the effectiveness of the [portalTimeout](https://hieromon.github.io/AutoConnect/apiconfig.html#portaltimeout), and the sketch control will go into the *loop()* function. Is this the use case you envision?

In any case, there is no AutoConnect settings that would directly satisfy your use case. It will always involve control on the sketch side.

**Reply by hjkoster400d** [reply]
'When nothing happens' <= means in this case, when the connection could not be restored with a known AP.
